src/main/scala/firrtl/passes/InferWidths.scala

```diff
@@ -45,200 +40,251 @@ import firrtl.WrappedExpression._
 
-       val ex = remove_cycle(n)(e_sub)
+      //println("Solving " + n + " => " + e)
+      //println("After Substitute: " + n + " => " + e_sub)
+      //println("==== SOLUTIONS TABLE (Post Substitute) ====")
+      //for (x <- f) println(x)
+      //println("=========================")
 
-       //println("After Remove Cycle: " + n + " => " + ex)
-       if (!self_rec(n,ex)) {
-         //println("Not rec!: " + n + " => " + ex)
-         //println("Adding [" + n + "=>" + ex + "] to Solutions Table")
-         o += n
-         f(n) = ex
-       }
-     }
-  
-     //println("Forward Solved Constraints")
-     //for (x <- f) println(x)
-  
-     //; Backwards Solve
-     val b = LinkedHashMap[String,Width]()
-     for (i <- 0 until o.size) {
-       val n = o(o.size - 1 - i)
-       /*
-       println("SOLVE BACK: [" + n + " => " + f(n) + "]")
-       println("==== SOLUTIONS TABLE ====")
-       for (x <- b) println(x)
-       println("=========================")
-       */
-       val ex = simplify(b_sub(b)(f(n)))
-       /*
-       println("BACK RETURN: [" + n + " => " + ex + "]")
-       */
-       b(n) = ex
-       /*
-       println("==== SOLUTIONS TABLE (Post backsolve) ====")
-       for (x <- b) println(x)
-       println("=========================")
-       */
-     }
-     b
+      val ex = remove_cycle(n)(e_sub)
+
+      //println("After Remove Cycle: " + n + " => " + ex)
+      if (!hasVarWidth(n)(ex)) {
+        //println("Not rec!: " + n + " => " + ex)
+        //println("Adding [" + n + "=>" + ex + "] to Solutions Table")
+        f(n) = ex
+        o += n
+      }
+    }
+ 
+    //println("Forward Solved Constraints")
+    //for (x <- f) println(x)
+ 
+    //; Backwards Solve
+    val b = LinkedHashMap[String, Width]()
+    for (n <- o) {
+      /*
+      println("SOLVE BACK: [" + n + " => " + f(n) + "]")
+      println("==== SOLUTIONS TABLE ====")
+      for (x <- b) println(x)
+      println("=========================")
+      */
+      val ex = simplify(b_sub(b)(f(n)))
+      /*
+      println("BACK RETURN: [" + n + " => " + ex + "]")
+      */
+      b(n) = ex
+      /*
+      println("==== SOLUTIONS TABLE (Post backsolve) ====")
+      for (x <- b) println(x)
+      println("=========================")
+      */
+    }
+    b
   }
      
-  def width_BANG (t:Type) : Width = {
-    (t) match {
-      case (t:UIntType) => t.width
-      case (t:SIntType) => t.width
-      case ClockType => IntWidth(1)
-      case (t) => error("No width!"); IntWidth(-1) } }
-  def width_BANG (e:Expression) : Width = width_BANG(e.tpe)
+  def run (c: Circuit): Circuit = {
+    val v = ArrayBuffer[WGeq]()
+
+    def get_constraints_t(t1: Type, t2: Type, f: Orientation): Seq[WGeq] = (t1,t2) match {
+      case (t1: UIntType, t2: UIntType) => Seq(WGeq(t1.width, t2.width))
+      case (t1: SIntType, t2: SIntType) => Seq(WGeq(t1.width, t2.width))
+      case (t1: BundleType, t2: BundleType) => t1.fields zip t2.fields flatMap {
```

Comment:
I'm ok with this change, but flatMap here might well be slower. We need to watch out for it.